rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala

```diff
@@ -0,0 +1,790 @@
+    )
+    deserializedNode shouldBe referenceNode
+    serializedNode shouldBe referenceSerialized
+  }
+
+  "decode wrong serialized data" should "be stopped with assertion error" in {
+    val wrongSerialized = createBV(
+      "0102FFFF00000000000000000000000000000010280FFFFFFFFFFFFFFFFFFFFFFFFFF"
+    )
+    try {
+      RadixTree.Codecs.decode(wrongSerialized)
+    } catch {
+      case ex: AssertionError =>
+        ex shouldBe a[AssertionError]
+        ex.getMessage shouldBe s"assertion failed: Error during deserialization: invalid data format"
+    }
+  }
+
+  "collisions in KVDB" should "be detected" in withImplAndStore { (impl, inMemoStore) =>
+    def copyBVToBuf(bv: ByteVector): ByteBuffer = {
+      val arr    = bv.toArray
+      val newBuf = ByteBuffer.allocateDirect(arr.length)
+      newBuf.put(arr).rewind()
+    }
+
+    val collisionKVPair = (copyBVToBuf(RadixTree.emptyRootHash.bytes), ByteVector(0x00))
+    val referenceErrorMessage = s"1 collisions in KVDB (first collision with key = " +
+      s"${RadixTree.emptyRootHash.bytes.toHex})."
+    for {
+      _   <- inMemoStore.put[ByteVector](Seq(collisionKVPair), copyBVToBuf)
+      _   = impl.saveNode(emptyNode)
+      err <- impl.commit.attempt
+
+      ex = err.left.get
+      _  = ex shouldBe a[RuntimeException]
+      _  = ex.getMessage shouldBe referenceErrorMessage
+    } yield ()
+  }
+
+  "tree with saveAndCommit" should "be built correctly and not create artefacts in KV - store" in withImplAndStore {
+    (impl, inMemoStore) =>
+      def createDeleteActions(keys: List[ByteVector]): List[DeleteAction] =
+        keys.map(key => DeleteAction(key.toSeq))
+
+      /* treeDataSet:
+            key      |   value
+        111122334455 | 0000...0001
+        11112233AABB | 0000...0002
+        1111AABBCC   | 0000...0003
+        33           | 0000...0004
+        FF0011       | 0000...0005
+        FF012222     | 0000...0006
+       */
+      val deleteActions = createDeleteActions(treeDataSet.map(_.rKey))
+
+      val referenceTree1 = Vector(
+        "TREE1: root =>",
+        "   [11]PTR: prefix = 11, ptr =>",
+        "      [22]PTR: prefix = 33, ptr =>",
+        "         [44]LEAF: prefix = 55, data = 0000...0001",
+        "         [AA]LEAF: prefix = BB, data = 0000...0002",
+        "      [AA]LEAF: prefix = BBCC, data = 0000...0003",
+        "   [33]LEAF: prefix = empty, data = 0000...0004",
+        "   [FF]PTR: prefix = empty, ptr =>",
+        "      [00]LEAF: prefix = 11, data = 0000...0005",
+        "      [01]LEAF: prefix = 2222, data = 0000...0006"
+      )
+      val referenceTree2 = Vector("TREE2: root =>")
+      for {
+        //  1  Build a tree according to the example in specification
+        rootNodeAndHash1Opt <- impl.saveAndCommit(RadixTree.emptyNode, referenceInsertActions)
+        (rootNode1, _)      = rootNodeAndHash1Opt.get
+
+        //    Get the tree for compare with reference
```

Comment:
Different number of spaces in comment lines. We should use single space after slashes. Please check this whole file.